### PR TITLE
feat:Updated Local Enquiry Report

### PR DESCRIPTION
**Pull request by ZakkiyathLulu**
##Feature Overview
Assignment of Enquiry Officer: Added a validation to ensure that an enquiry officer is selected before progressing the workflow to "Assign to Enquiry Officer" state.
Manager Role Validation: Ensured that only managers can transition the workflow from "Draft" to "Assign to Enquiry Officer."
Start and Completion Date Automation: Automatically sets the start and  completion date based on workflow states.
Automated Workflow and Status Updates:

Workflow restrictions allow only authorized users (e.g., HR manager,Enquiry Manager and Enquiry officer) to transition certain states.

## Solution Summary
Modified workflow states:State Transitions:
Draft->Assign to Admin->Assigned to Admin(HR Manager)
Assigned to Admin->Assign to Enquiry Officer->Assigned to Enquiry Officer(Enquiry Manager)--->
Assigned to Enquiry Officer->Start Enquiry->Enquiry on Progress(Enquiry Officer)------------->
Enquiry on Progress->Enquiry Completed->Pending Approval(Enquiry Officer)
Assigned to Enquiry Officer->Enquiry Completed->Pending Approval(Enquiry Officer)
Pending Approval->Approve->Approved(HR Manager)
Pending Approval->Reject->Rejected(HR Manager)

Added validation logic in validate() function of Local Enquiry Report doctype to check the selection of an enquiry officer.

The Local Enquiry Report document is assigned for Enquiry manger from draft state and He thenreassigns it to Enquiry Officer.This time assignment with Enquiry Manager is removed.
Automated Date Management: Set Enquiry Start and Completion dates automatically when  corresponding workflow states  are triggered(Start Enquiry  and enquiry Completed).

## Files Affected
beams/beams/custom_scripts/local_enquiry_report/local_enquiry_report.py

##Screenshots
image
imageimage
image
image(since condition given in workflow the action button will show only if the field enquiry officer is selected)
image
image
image
image
image
image
image(start date populated (read only ))
image(completion date set)
image
image
image(workflow)
(workflow)

## Testing and Dataset

Tested with Existing Data: Verified workflow and automated status changes on existing reports.
Tested with New Data: Created new reports to ensure correct assignment and validation of officers and managers.

## Behavior Changes
 No other features affected by this change.

## Patch Required
       No

## Was this feature tested on the browsers?
  - Chrome
  - Mozilla Firefox

### CODE REVIEW

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.js [hunk 1]
```diff
@@ -1,7 +1,27 @@
-// Copyright (c) 2024, efeone and contributors
+// Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and contributors
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.js:1]: don't change this

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 2]
```diff
@@ -66,21 +133,158 @@ def set_status_to_overdue():
             # Check if 'Enquiry Completion Date' is not set or is later than 'Expected Completion Date'
             if not enquiry.enquiry_completion_date or getdate(enquiry.enquiry_completion_date) > getdate(enquiry.expected_completion_date):
                 frappe.db.set_value('Local Enquiry Report', enquiry.name, 'status', 'Overdue')
+                frappe.db.commit()  # Commit the changes to the database
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:136]: is this necessary?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 3]
```diff
@@ -66,21 +133,158 @@ def set_status_to_overdue():
...
 
-        frappe.db.commit()
+    frappe.msgprint("Overdue status has been updated for applicable Local Enquiry Reports.")
 
 @frappe.whitelist()
 def update_job_applicant_status(local_enquiry_report):
+
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:142]: not required

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 4]
```diff
@@ -66,21 +133,158 @@ def set_status_to_overdue():
...
+        manager_id = managers[0].parent
+        add_assign({
+            "assign_to": [manager_id],
+            "doctype": "Local Enquiry Report",
+            "name": doc.name,
+            "description": "Please review and assign to an Enquiry Officer."
+        })
+
+        frappe.msgprint(_("Enquiry Manager  has been Assigned."))
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:178]: remove double space

File: beams/hooks.py [hunk 5]
```diff
@@ -203,7 +203,7 @@
     },
     "Interview": {
         "on_update": "beams.beams.custom_scripts.job_applicant.job_applicant.update_applicant_interview_round"
-    }
+        }
```
Comment (MhmdSinanKT) [inline beams/hooks.py:206]: unnecessary changes

**Review by MhmdSinanKT — CHANGES_REQUESTED** [review]

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.js [hunk 6]
```diff
@@ -1,7 +1,27 @@
...
-// });
+frappe.ui.form.on('Local Enquiry Report', {
+  refresh: function (frm) {
+      // Set fields to read-only if they already have values
+      if (frm.doc.enquiry_start_date) {
+          frm.set_df_property("enquiry_start_date", "read_only", 1);
+      }
+      if (frm.doc.enquiry_completion_date) {
+          frm.set_df_property("enquiry_completion_date", "read_only", 1);
+      }
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.js:12]: use set only once setting in doctype

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 7]
```diff
@@ -1,16 +1,28 @@
-# Copyright (c) 2024, efeone and contributors
+# Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and contributors
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:1]: undo this

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 8]
```diff
@@ -1,16 +1,28 @@
...
+from frappe.utils import getdate, add_days, today, nowdate
+from frappe import _
+from frappe.desk.form.assign_to import add as add_assign
+from frappe.utils.user import get_users_with_role
 
 class LocalEnquiryReport(Document):
+
     def validate(self):
-        self.information_required()
+        self.information_required()  # Assuming you need to call this once
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:14]: unnecessary comment

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 9]
```diff
@@ -1,16 +1,28 @@
...
-        
+
+        """Ensure enquiry officer selection based on workflow state."""
+        # Check if the workflow state is 'Assign to Enquiry Officer'
+        if self.workflow_state == 'Assign to Enquiry Officer':
+            if not self.enquiry_officer:
+                frappe.msgprint(_("Please select an enquiry officer before proceeding."))
+            else:
+                # If the enquiry officer is selected, update the field
+                frappe.db.set_value("Local Enquiry Report", self.name, "enquiry_officer", self.enquiry_officer)
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:24]: What is this line supposed to do? this line is simmply setting the same value in the same field

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 10]
```diff
@@ -24,18 +36,45 @@ def information_required(self):
...
-                missing_fields.append("Information given by Designation")
+            # Ensure enquiry officer selection based on workflow state.
+            if self.workflow_state == 'Assigned to Enquiry Officer':
+                if not self.enquiry_officer:
+                    frappe.msgprint(_("Please select an enquiry officer before proceeding."))
+                    # Optionally, raise an exception to prevent saving the document if no officer is selected
+                    raise frappe.ValidationError(_("Enquiry officer must be selected when the state is 'Assigned to Admin'."))
+                else:
+                    # If the enquiry officer is selected, update the field
+                    frappe.db.set_value("Local Enquiry Report", self.name, "enquiry_officer", self.enquiry_officer)
```
Comment (MhmdSinanKT) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:47]: What is this line supposed to do? this line is simmply setting the same value in the same field

File: beams/public/js/job_applicant.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+              const applicant_skills = frm.doc.skill_proficiency.map(skill => skill.skill);
+              const missing_skills = required_skills.filter(skill => !applicant_skills.includes(skill));
+
+              if (missing_skills.length > 0) {
+                  const required_skills_list = required_skills.join(", ");
+                  frappe.msgprint(__("The Applicant does not meet the Required skills: {0}.").format(required_skills_list));
+                  frappe.validated = false;
+                  return;
+              }
+          }
```
Comment (MhmdSinanKT) [inline beams/public/js/job_applicant.js:41]: do this in python, client side validations should be reactive

File: beams/public/js/job_requisition.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+              __("Create Job Opening"),
+              () => {
+                // Open a mapped document to create a Job Opening
+                  frappe.model.open_mapped_doc({
+                      method: "beams.beams.custom_scripts.job_requisition.job_requisition.make_job_opening",
+                      source_name: frm.doc.name,  // Pass the current Job Requisition ID
+                  });
+              },
+              __("Create"),
+              "btn-secondary"  // Set as a secondary button
```
Comment (MhmdSinanKT) [inline beams/public/js/job_requisition.js:33]: is this necessary?

File: beams/public/js/job_requisition.js [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+                                  );
+                              },
+                          });
+                      },
+                      __("Associate Job Opening"),
+                      __("Submit")
+                  );
+              },
+              __("Create"),
+              "btn-secondary"  // Set as a secondary button
```
Comment (MhmdSinanKT) [inline beams/public/js/job_requisition.js:93]: is this necessary?

File: beams/public/js/job_requisition.js [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+// Rename Actions Button  as 'Create'  not primary action button
+     $(document).ready(function () {
+          // Find the specific custom "Actions" button dropdown using the parent or group
+          var dropdownActions = frm.page.inner_toolbar
+              .find('button:contains("Actions")')
+              .first();
+
+          // Rename this specific "Actions" button to "Create"
+          if (dropdownActions.length) {
+              dropdownActions.text("Create");
```
Comment (MhmdSinanKT) [inline beams/public/js/job_requisition.js:138]: use snake_case

**Review by MhmdSinanKT — CHANGES_REQUESTED** [review]

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 15]
```diff
@@ -3,14 +3,23 @@
...
+
     def validate(self):
         self.information_required()
         self.set_expected_completion_date()
-        self.information_required()
-        
+
+        """Ensure enquiry officer selection based on workflow state."""
+        # Check if the workflow state is 'Assign to Enquiry Officer'
+        if self.workflow_state == 'Assigned to Admin':
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:19]: Check the workflow_sate condition with in line documentation

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 16]
```diff
@@ -24,18 +33,45 @@ def information_required(self):
...
                 frappe.throw("Please provide 'Information given by' and 'Information given by Designation' before completing the report.")
 
             elif missing_fields:
                 frappe.throw(f"Please provide '{', '.join(missing_fields)}' before completing the report.")
 
+    def on_update(doc):
+        """Handle workflow transitions upon document update."""
+        try:
+            if isinstance(doc, str):
+                doc = frappe.get_doc('Local Enquiry Report', doc)
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:56]: Why this is required?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 17]
```diff
@@ -66,21 +102,156 @@ def set_status_to_overdue():
...
+        managers = frappe.get_all(
+            "Has Role",
+            filters={"role": "Enquiry Manager", "parenttype": "User"},
+            fields=["parent"]
+        )
+
+        if not managers:
+            frappe.throw(_("No users found with the 'Enquiry Manager' role."))
+
+        manager_id = managers[0].parent
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:139]: User default frappe methods to get user with role

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 18]
```diff
@@ -66,21 +102,156 @@ def set_status_to_overdue():
...
+
+    except Exception as e:
+        frappe.throw(str(e))
+
+@frappe.whitelist()
+def assign_to_enquiry_officer(doc):
+    """Assign LER to the selected Enquiry Officer."""
+    try:
+        if not isinstance(doc, Document):
+            doc = frappe.get_doc("Local Enquiry Report", doc)
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:157]: Why this is required?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 19]
```diff
@@ -66,21 +102,156 @@ def set_status_to_overdue():
...
+        todo = frappe.get_value(
+            "ToDo",
+            {
+                "reference_type": "Local Enquiry Report",
+                "reference_name": doc.name,
+                "description": manager_assignment_description,
+                "status": "Open"
+            },
+            "name"
+        )
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:203]: Use builtin method to get assignments and Close

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 20]
```diff
@@ -66,21 +102,156 @@ def set_status_to_overdue():
...
+@frappe.whitelist()
+def set_enquiry_start_date(doc):
+    """
+    This function checks if the 'enquiry_start_date' is empty or not. If it is not set,
+    the function assigns the current date to the field and saves the document.
+    """
+
+    if not doc.enquiry_start_date:
+        doc.enquiry_start_date = nowdate()
+        doc.save(ignore_permissions=True)
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:221]: Is this tested properly? Why doc.save() is required on workflow actions?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 21]
```diff
@@ -66,21 +102,156 @@ def set_status_to_overdue():
...
+
+@frappe.whitelist()
+def set_enquiry_completion_date(doc):
+    """
+    This function checks if the 'enquiry_completion_date' is empty or not. If it is not set,
+    the function assigns the current date to the field and saves the document.
+    """
+    if not doc.enquiry_completion_date:
+        doc.enquiry_completion_date = nowdate()
+        doc.save(ignore_permissions=True)
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:232]: Is this tested properly? Why doc.save() is required on workflow actions?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 22]
```diff
@@ -66,21 +102,156 @@ def set_status_to_overdue():
...
+        FROM `tabEmployee` emp
+        JOIN `tabUser` u ON u.name = emp.user_id
+        JOIN `tabHas Role` hr ON hr.parent = u.name
+        WHERE hr.role = %s
+        AND emp.status = 'Active'
+        AND (emp.name LIKE %s OR emp.employee_name LIKE %s)
+        LIMIT %s, %s
+        """,
+        (role_filter, f"%{txt}%", f"%{txt}%", start, page_len),
+        as_list=True
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:251]: Please restructure the Query in a readable format

**Review by SherinKR — CHANGES_REQUESTED** [review]

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 23]
```diff
@@ -3,14 +3,24 @@
...
+
     def validate(self):
         self.information_required()
         self.set_expected_completion_date()
-        self.information_required()
-        
+
+        """Ensure enquiry officer selection based on workflow state."""
+        # Ensure enquiry officer selection based on workflow state.
+        if self.workflow_state == 'Assign to Enquiry Officer':
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:20]: Is workflow_state name is correct?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 24]
```diff
@@ -24,18 +34,42 @@ def information_required(self):
                 self.information_collected_by = frappe.session.user
 
             missing_fields = []
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:36]: Why missing_fields is used?

File: beams/beams/doctype/local_enquiry_report/local_enquiry_report.py [hunk 25]
```diff
@@ -66,21 +100,196 @@ def set_status_to_overdue():
...
+    except Exception as e:
+        frappe.throw(str(e))
+
+
+@frappe.whitelist()
+def assign_to_enquiry_officer(doc):
+    """Assign LER to the selected Enquiry Officer."""
+    try:
+        # Fetch user ID of the enquiry officer
+        user_id = frappe.db.get_value("Employee", {"name": doc.enquiry_officer}, "user_id")
```
Comment (SherinKR) [inline beams/beams/doctype/local_enquiry_report/local_enquiry_report.py:156]: Check existence of the document before using get_value

**Review by SherinKR — CHANGES_REQUESTED** [review]

**Pull request merged**